src/Trinity.Core/Trinity/Runtime/TrinityC.cs

```diff
@@ -21,43 +23,71 @@ internal static class TrinityC
 #else
         internal const string AssemblyName = "Trinity.C.dll";
 #endif
-        static TrinityC()
+        private static object s_initlock = new object();
+        private static bool   s_initialized = false;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Init()

```

Comment:
It is a little bit messy here because of different init logic for CORECLR and non-CORECLR platforms. I would suggest split it into different methods.